pkg/apis/core/v1beta1/helper/helper.go

```diff
@@ -682,45 +658,125 @@ func FindPrimaryDNSProvider(providers []gardencorev1beta1.DNSProvider) *gardenco
+
+	// Try to find non-deprecated version first
+	qualifyingVersionFound, latestNonDeprecatedImageVersion, err := GetLatestQualifyingVersion(versions, append(versionPredicates, FilterDeprecatedVersion())...)
 	if err != nil {
 		return false, "", err
 	}
-	// latest version cannot be found. Do not return an error, but allow for minor upgrade if Shoot's machine image version is expired.
+	if qualifyingVersionFound {
+		return true, latestNonDeprecatedImageVersion.Version, nil
+	}
+
+	// otherwise, also consider deprecated versions
+	qualifyingVersionFound, latestVersion, err := GetLatestQualifyingVersion(versions, versionPredicates...)
+	if err != nil {
+		return false, "", err
+	}
+	// latest version cannot be found. Do not return an error, but allow for forceful upgrade if Shoot's version is expired.
 	if !qualifyingVersionFound {
 		return false, "", nil
 	}
 
 	return true, latestVersion.Version, nil
 }
 
-// GetKubernetesVersionForMinorUpdate finds a Kubernetes version in the <cloudProfile> that qualifies for a Kubernetes minor level update given a <currentVersion>.
-// A qualifying version is a non-preview version having the minor version increased by exactly one version.
-// In case the consecutive minor version has only expired versions, picks the latest expired version (will do another minor update during the next maintenance time)
+// GetVersionForForcefulUpdateToConsecutiveMinor finds a version from a slice of expirable versions that qualifies for a minor level update given a <currentVersion>.
+// A qualifying version is a non-preview version having the minor version increased by exactly one version (required for Kubernetes version upgrades).
+// In case the consecutive minor version has only expired versions, picks the latest expired version (will try another update during the next maintenance time)
 // If a version can be found, returns true and the qualifying patch version of the next minor version.
-// In case it does not find a version, it returns false.
-func GetKubernetesVersionForMinorUpdate(cloudProfile *gardencorev1beta1.CloudProfile, currentVersion string) (bool, string, error) {
+// In case it does not find a version, it returns false and an empty string.
+func GetVersionForForcefulUpdateToConsecutiveMinor(versions []gardencorev1beta1.ExpirableVersion, currentVersion string) (bool, string, error) {
 	currentSemVerVersion, err := semver.NewVersion(currentVersion)
 	if err != nil {
 		return false, "", err
 	}
 
-	qualifyingVersionFound, latestVersion, err := GetLatestQualifyingVersion(cloudProfile.Spec.Kubernetes.Versions, FilterNonConsecutiveMinorVersion(*currentSemVerVersion), FilterSameVersion(*currentSemVerVersion), FilterExpiredVersion())
+	// filters out any version that does not have minor version +1
+	predicates := []VersionPredicate{FilterDifferentMajorVersion(*currentSemVerVersion), FilterNonConsecutiveMinorVersion(*currentSemVerVersion)}
```

Comment:
Filter for same version?
```suggestion
	predicates := []VersionPredicate{FilterDifferentMajorVersion(*currentSemVerVersion), FilterNonConsecutiveMinorVersion(*currentSemVerVersion), FilterSameVersion(*currentSemVerVersion)}
```

Reply:
That's already included in predicate `FilterNonConsecutiveMinorVersion` as the same version as the current version cannot satisfy `current_version = version_to_compare +1`, hence is filtered out.